tests/integration/test_client_http.py

```diff
@@ -0,0 +1,203 @@
+        config = Config(read_timeout=0.1, retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        class FakeEC2(SimpleHandler):
+            msg = b'<response/>'
+            def get_length(self):
+                return len(self.msg)
+
+            def get_body(self):
+                time.sleep(1)
+                return self.msg
+
+        try:
+            with background(run_server, args=(FakeEC2, self.port)):
+                client.describe_regions()
+        except ReadTimeoutError:
+            # Note: This was vendored requests ConnectionError & ReadTimeout
+            pass
+        except BackgroundTaskFailed:
+            self.fail('Fake EC2 service was not called.')
+        else:
+            self.fail('Excepted exception was not thrown')
+
+    def test_connect_timeout_exception(self):
+        config = Config(connect_timeout=0.2, retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        def no_accept_server():
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            sock.bind(('', self.port))
+            time.sleep(4)
+            sock.close()
+
+        try:
+            with background(no_accept_server):
+                time.sleep(2)
+                client.describe_regions()
+        except ConnectTimeoutError:
+            pass
+        except BackgroundTaskFailed:
+            self.fail('Server failed to exit in a timely manner.')
+        else:
+            self.fail('Excepted exception was not thrown')
+
+    def test_invalid_host_gaierror(self):
+        config = Config(retries={'max_attempts': 0})
+        endpoint = 'https://ec2.us-weast-1.amazonaws.com/'
+        client = self.session.create_client('ec2', endpoint_url=endpoint,
+                                            config=config)
+        try:
+            client.describe_regions()
+        except EndpointConnectionError:
+            pass
+
+    def test_bad_status_line(self):
+        config = Config(retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        class BadStatusHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_POST(self):
+                self.wfile.write(b'garbage')
+
+        try:
+            with background(run_server, args=(BadStatusHandler, self.port)):
+                client.describe_regions()
+        except ConnectionClosedError:
+            pass
+        except BackgroundTaskFailed:
+            self.fail('Fake EC2 service was not called.')
+        else:
+            self.fail('Excepted exception was not thrown')
+
+
+def random_port():
```

Comment:
I would suggest doing what we did for chalice where we have the socket module find an unused port for us.  This function has a small chance of failing, but it'd be nice to have something that we know will always work: https://github.com/aws/chalice/blob/master/tests/functional/test_local.py#L95

Reply:
Done